dwl.c

```diff
@@ -1226,16 +1242,29 @@ rendermon(struct wl_listener *listener, void *data)
 	struct timespec now;
 	clock_gettime(CLOCK_MONOTONIC, &now);
 
+	/* Do not render if clients have uncommitted changes. */
+	wl_list_for_each(c, &stack, slink)
+       	{
+		if (c->dirty)
+	       	{
+			wlr_surface_send_frame_done(c->xdg_surface->surface, &now);
+			render = 0;
+		}
+	}
+
 	/* wlr_output_attach_render makes the OpenGL context current. */
 	if (!wlr_output_attach_render(m->wlr_output, NULL))
 		return;
 
 	/* Begin the renderer (calls glViewport and some other GL sanity checks) */
 	wlr_renderer_begin(drw, m->wlr_output->width, m->wlr_output->height);
-	wlr_renderer_clear(drw, rootcolor);
 
-	renderclients(m, &now);
-	renderindependents(output, &now);
+	if (render) {
+		wlr_renderer_clear(drw, rootcolor);
+
+		renderclients(m, &now);
+		renderindependents(output, &now);
+	}
```

Comment:
It appears necessary to call render begin / end, otherwise the loop calling render seems to... stop.

Reply:
Hmm. The frame_done events made it work for me, but I'd really like to know where the hangup is so I know we're doing the right thing here.

(Also, brace style on the `rendermon` changes doesn't match the rest of the code.)

Reply:
> Hmm. The frame_done events made it work for me, but I'd really like to know where the hangup is so I know we're doing the right thing here.

I'm not sure it was actually working: I tried the atomic-arrange branch, adding some log lines to rendermon. It stops being called the first time we set defer = 1.

Digging into wlroots / sway to see what is going on.

Reply:
sway has set up its damage detection in such a way that its render callback `damage_handle_frame` is only invoked when there are changes to render. `wlr_renderer_begin` / `end` are not called during that callback; instead they are called as part of sway's transaction handling mechanism, as a result of user input. This explains nothing.

wlroots invokes rendermon in response to wayland DRM events:
```
#0  rendermon (listener=0x555555f1a888, data=0x555555f19f40) at dwl.c:1248
#1  0x00007ffff7f7fc08 in wlr_signal_emit_safe (signal=0x555555f1a0c8, data=0x555555f19f40) at ../util/signal.c:29
#2  0x00007ffff7f6da6a in wlr_output_send_frame (output=0x555555f19f40) at ../types/wlr_output.c:661
#3  0x00007ffff7f28452 in page_flip_handler (fd=9, seq=689548, tv_sec=11553, tv_usec=356413, crtc_id=51, data=0x55555558b2a0)
    at ../backend/drm/drm.c:1527
#4  0x00007ffff7c47477 in drmHandleEvent () at /usr/lib/libdrm.so.2
#5  0x00007ffff7f284d7 in handle_drm_event (fd=9, mask=1, data=0x0) at ../backend/drm/drm.c:1537
#6  0x00007ffff7ec8faa in wl_event_loop_dispatch () at /usr/lib/libwayland-server.so.0
#7  0x00007ffff7ec74e7 in wl_display_run () at /usr/lib/libwayland-server.so.0
#8  0x000055555555c729 in run (startup_cmd=0x0) at dwl.c:1371
#9  0x000055555555db00 in main (argc=1, argv=0x7fffffffe838) at dwl.c:1874
```
An apparent side effect of not calling `wlr_renderer_begin` / `end` is these events not being generated or being filtered out.

I have not answered anything here, however this should all become clear as we implement damage detection.